### ISSUE TITLE
[useDirection] Eager hook is called on every state change.

### ISSUE THREAD

**Issue by steveruizok** [radix-ui/primitives#749]
### Current Behavior

Several components (such as the [`DropdownMenu`](https://github.com/radix-ui/primitives/blob/main/packages/react/dropdown-menu/src/DropdownMenu.tsx) component) use the [`useDirection`](https://github.com/radix-ui/primitives/blob/main/packages/react/use-direction/src/useDirection.tsx#L15) hook. 

If this hook does not receive a `dir` prop, the hook [reads the element's computed styles](https://github.com/radix-ui/primitives/blob/main/packages/react/use-direction/src/useDirection.tsx#L1) by pulling the `direction` property from the element's computed style. 

This can be expensive. As none of the elements provide a default for this prop, if used on an element that might changes very often for whatever reason, this will cause the hook to run on every frame. This was happening to me: see a before and after before I went source-code spelunking.

Before (with `<Dropdown.Root>`): [Video Link](
https://www.dropbox.com/s/dbl3huzumayhifz/Kapture%202021-06-27%20at%2019.18.12.mp4?dl=0)

After (with `<Dropdown.Root dir="ltr">`): [Video Link](
https://www.dropbox.com/s/u51np3gdflkpijo/Kapture%202021-06-27%20at%2019.25.02.mp4?dl=0)

### Expected behavior

If no property is provided, the hook should stick with its first result rather than repeating the expensive job of getting the element's computed styles.

### Reproducible example

Extremely contrived
https://codesandbox.io/s/nervous-jackson-plg87?file=/src/App.tsx

### Suggested solution

At the risk of being a touch imperialist, I'd expect `ltr` to be the default for the `dir` prop. 

If you do want to read the element itself to get the direction, then I'd expect that the hook did not refresh when the element changes. Specifically, I'd suggest that `element` be dropped from [this dependency array](https://github.com/radix-ui/primitives/blob/main/packages/react/use-direction/src/useDirection.tsx#L19).

Otherwise, if you want to foreground the `ltr` or `rtl` choice, make it a required prop.

If none of the above are possible, I'd expect to see this flagged as a potential performance issue when leaving it undefined for components that may change often.

### Your environment

| Software         | Name(s) | Version |
| ---------------- | ------- | ------- |
| Radix Package(s) |   @radix-ui/react-dropdown      |   0.0.22      |
| React            | n/a     |   Latest / see sandbox      |
| Browser          |         |      `   |
| Assistive tech   |         |    `     |
| Node             | n/a     |     `    |
| npm/yarn         |         |     `    |
| Operating System |         |    `     |

**Comment by benoitgrelard** [comment]
Hey @steveruizok, thanks for that.

Before we can look much further, can you help me understand why the element itself might change often? Which use-case/primitive did you find that to be the case?

**Comment by steveruizok** [comment]
I believe it was the DropdownMemu that caused it, though I'm not sure which. I have a few menus that are shown conditionally in the app, depending on how many items are selected, but none of those options would have been changing during a drag session. If I had to guess now, I'd say it was the style panel dropdown at the top right, which currently updates whenever selected elements change. 

I'll look back at the code (perhaps it's indicative of a different issue). It seemed like anything that caused the element to render, like an unmemoized callback or object prop, would also cause the hook to run.

**Comment by benoitgrelard** [comment]
I'm asking because although the hook will run on every render, the expensive part `getComputedStyles` will only run again if the element changes. This is ok because the style declaration you get from `getComputedStyles` is a live declaration, meaning that it is always up to date without having to incur another call to `getComputedStyles`.

So the hook code basically can be summed up as:

- run `getComputedStyles` if the element or the `dir` prop itself changes (most likely would be only once alltogether)
- if the `dir` prop isn't defined directly (inherited) we continuously read it from the cached live computedStyles in a raf loop

Reading it shouldn't be expensive as far as I know, so that's why I'm wondering if it's something else you're seeing here.

**Comment by jjenzz** [comment]
As Benoit has already explained, `getComputedStyle` will only be called once as the `element` should not change on re-render. I've done some tests locally with your reproduction code and also confirmed this there. 

I'm getting the following Call Tree results:

| Screenshot | Results |
| - | - |
| ![image](https://user-images.githubusercontent.com/175330/123877038-5e62eb80-d934-11eb-828f-1e5530c15a22.png) | The `getStorybookSetCountLoop` (the loop from your codesandbox that is updating the count and triggering the re-renders) is 640ms of the total time while our `getDirection` loop is only 8.5ms. |

If I remove all of our direction logic so that the effect is simply calling an empty `requestAnimationFrame` loop, there isn't a noticeable improvement in perf so it seems the real issue could be somewhere else.

Looking at the code you have written to reproduce the issue, are you suggesting that the drag logic in your app is updating some React state over and over again with the position of the shapes you are rendering on screen? It's usually not recommended to trigger lots of re-renders for things like that so if you are then improving perf there would probably help.

**Comment by jjenzz** [comment]
Closing this for now because it's unlikely that the rAF loop is the source of the performance issues. Happy to reopen this thought of course if some use-cases can be provided that show otherwise.

**Comment by dwilt** [comment]
@steveruizok thank you so much for being a touch imperialist - you saved me and my dev team lots of digging around into a very large performance problem in our app. We have _dozens_ of instances of the Radix Dropdown menu in our app:

![2021-10-11 16 04 22](https://user-images.githubusercontent.com/1065539/136849325-511dca62-faea-42c4-95bf-3352717c5022.gif)

I took the below performance recording while doing _nothing_ in our web app. I just hit record. I didn't even move the mouse over the application - it never left the dev tools. You can see hundreds of rAF calls clogging up the JS thread running on every frame.

![2021-10-11 16 02 01](https://user-images.githubusercontent.com/1065539/136849040-71939cba-8894-4bc3-b6a6-e17c13dc6105.gif)

@jjenzz I don't think [your performance measuring](https://github.com/radix-ui/primitives/issues/749#issuecomment-870987883) was really fair because you only have 1 instance of the dropdown menu on the page and you weren't running the CPU at 6x slowdown. [It's a known good practice](https://kentcdodds.com/blog/profile-a-react-app-for-performance#pitfall-profiling-with-a-fast-computer) to be using 6x slowdown on the CPU when doing performance testing because you (as a developer) likely have a much faster computer than a lot of regular users.

We applied the `dir="ltr"` to the `Root` component and it did fix the performance issue. Agree with @steveruizok that this should either be set as a default in the hook. As a library who probably strives to be used by large companies in production-grade applications, I'm surprised that this issue was closed saying that this performance issue wasn't important.

**Comment by benoitgrelard** [comment]
Hi @dwilt,

Thanks for giving details of your experience with this.
I still stand by my previous comment though (https://github.com/radix-ui/primitives/issues/749#issuecomment-869544504).

> I took the below performance recording while doing nothing in our web app. I just hit record. I didn't even move the mouse over the application - it never left the dev tools. You can see hundreds of rAF calls clogging up the JS thread running on every frame.

There's a raf loop going on, so of course you will see things happening in the trace. If you have a raf loop running, even with no code inside, you will still see something in the trace. If you look carefully at your trace, you can see that each even is in fact in the order of μs (or at the worst below 1ms).

> @jjenzz I don't think your performance measuring was really fair because you only have 1 instance of the dropdown menu on the page and you weren't running the CPU at 6x slowdown. It's a known good practice to be using 6x slowdown on the CPU when doing performance testing because you (as a developer) likely have a much faster computer than a lot of regular users.

I also ran my own tests (we have stories with hundreds of instances of menus) with 6x CPU slowdown and I see the same results.

So I would be surprised if it's creating any performance issues. Also from what I can see in the gif where you're using your app, there doesn't seem to be any issues (that I can see). Similarly to my original response to Steve, I'm more interested in understanding how you got to the conclusion, what you observed, etc.

> I'm surprised that this issue was closed saying that this performance issue wasn't important.

Unless there's a real concern about performance, we probably won't make a change here. Especially given there's a very simple escape hatch to opt out of the whole "watch for inherited direction" mechanics (the raf loop).

Perhaps, you have other ideas as to how we could reliably detect a change in inherited `dir` without using a raf loop?

I can think of ways we could "mitigate" things by running it only where explicitly needed (for example we may get away with only running it when the menu is opened, as we probably don't explicitly need the dir for the trigger). But again, we generally only optimise things when we actually observe that there's an issue.

I hope this helps, and if you do have more details about it being an actual perf issue, we would love to understand and help resolve it.

**Comment by dwilt** [comment]
@benoitgrelard Take a look at [this video too](https://www.loom.com/share/bbb72a20832f499e9f7039f3ce0eb01b). I think this is a clue as to what might be going on.

**Comment by jjenzz** [comment]
I've just noticed that `RadioGroup`, `RovingFocusGroup`, `Tabs`, `ToggleGroup` and `Toolbar` all set a default value for their `dir` prop to `ltr` so this hook wouldn't be working for those anyway 🙈

I think the provider idea avoids the implicit nature of all of this "inheritence" idea so I'm more and more in favour of it.

**Comment by benoitgrelard** [comment]
Yeah we didn't take the time yet to apply that hook everywhere, so I guess when we do review that solution, if we go for the `Provider`, we'll update all primitives to read that context.

**Comment by steveruizok** [comment]
I'm still seeing this one on `0.1.1`, specifically in the `DropdownMenu`. 

I've let this ZoomMenu component update on each mouse move. When the `dir` prop is left undefined, the `useDirection` hook is running `getComputedStyles` on each update.

**Comment by jjenzz** [comment]
> I've let this ZoomMenu component update on each mouse move. When the `dir` prop is left undefined, the `useDirection` hook is running `getComputedStyles` on each update.

@steveruizok  It shouldn't be calling `getComputedStyle` on each mouse move, but it will be running the `requestAnimationFrame` loop. The `requestAnimationFrame` loop is [reading a property from a persisted object in state](https://github.com/radix-ui/primitives/blob/main/packages/react/use-direction/src/useDirection.tsx#L23), it doesn't call `getComputedStyle`. 

The only time we call `getComputedStyle` is on initial render and if the node changes so if you believe it is being called multiple times, it might be worth checking if your `DropdownMenu` is getting re-instantiated lots (e.g. `key` prop changing or remounting).

> The docs show a default for `dir` on the DropdownMenu root, but I'm pretty sure that this isn't the case. Was it meant to have that default?

Yep, this looks like a mistake in the docs 👍 

We haven't worked on the proposed provider solution yet but when we do we will straighten out the docs to reflect the change. Feel free to submit a docs PR in the meantime though so that they're more accurate till then if you like.

**Comment by ColmanHumphrey** [comment]
Ah we got caught by this too, and indeed setting the direction solved. I was seeing something like 6% CPU just sitting on the open tab according to Chrome's task manager, vs 0.1 after setting the direction (but I am not that knowledgeable about these numbers - maybe this isn't meaningful). We're on version 0.1.4, and I think with two or three dropdowns on the screen. I don't think we noticed user experience issues though

**Comment by steveruizok** [comment]
Ran into this one again on a new project. 

Without explicit `dir` prop:

<img width="1230" alt="image" src="https://user-images.githubusercontent.com/23072548/153396260-7df193b9-0ddd-4165-a550-aa02862d01f6.png">

With explicit `dir` prop:

<img width="1177" alt="image" src="https://user-images.githubusercontent.com/23072548/153396158-839069cf-c150-4e42-8a57-ec3a66bb0755.png">

This is just sitting idle, no interactions or anything.

**Comment by ScottSavarie** [comment]
> Ran into this one again on a new project.
> 
> Without explicit `dir` prop:
> 
> <img alt="image" width="1230" src="https://user-images.githubusercontent.com/23072548/153396260-7df193b9-0ddd-4165-a550-aa02862d01f6.png">
> 
> With explicit `dir` prop:
> 
> <img alt="image" width="1177" src="https://user-images.githubusercontent.com/23072548/153396158-839069cf-c150-4e42-8a57-ec3a66bb0755.png">
> 
> This is just sitting idle, no interactions or anything.

+1 over here. Could a quick fix be having `dir` set to `ltr` by default?

**Comment by latin-1** [comment]
@jjenzz 
1. `getComputedStyle(...)` itself is extremely fast since it only return a proxied **live** object, the expensive computing step will happen each time you get a property on it. Storing it inside a state is almost a no-op.
2. You should do layout related stuffs inside `useLayoutEffect` instead to avoid a DOM re-render in some cases. It also applies to other hooks like `useRect`.

Anyway, doing something that may cause style re-calculation and/or reflow in every frame is really a bad idea and should be avoided in most cases.

**Comment by jjenzz** [comment]
> Anyway, doing something that may cause style re-calculation and/or reflow in every frame is really a bad idea and should be avoided in most cases.

@latin-1 the team is very aware of this which is why radix generally tries to keep it to a minimum, but without premature optimisation. often times, the perf issues can be product architecture related (rendering too many things on the screen, updating state too frequently etc.) but if folks raise demonstrable concerns the team is always open to discussing them.

this ticket is being solved by removing inheritance from `useDirection` altogether and supplying a provider instead. please create a separate ticket for any `useRect` issues you're experiencing.

**Comment by JCMais** [comment]
When should we expect a new release with this change? (in case there is an ETA)

**Comment by benoitgrelard** [comment]
Hi @JCMais you can follow the current milestone here: https://github.com/radix-ui/primitives/milestones
